typing/jkind.ml

```diff
@@ -2278,32 +2283,207 @@ let for_boxed_variant cstrs =
+            match cstr.cd_args with
+            | Cstr_tuple args ->
+              List.fold_left
+                (fun (tys, ms) arg ->
+                  arg.ca_type :: tys, arg.ca_modalities :: ms)
+                (cstr_arg_tys, cstr_arg_modalities)
+                args
+            | Cstr_record lbls ->
+              List.fold_left
+                (fun (tys, ms) lbl ->
+                  lbl.ld_type :: tys, lbl.ld_modalities :: ms)
+                (cstr_arg_tys, cstr_arg_modalities)
+                lbls
+          in
+          (* Note: we're using polymorphic variants here to fake labeled tuples; we
+             can replace this with labeled tuples once we can build with a compiler
+             that supports those. *)
+          let `Args ret_args, `Params params, seen =
+            match cstr.cd_res with
+            | None when not has_gadt_constructors ->
+              `Args ret_args, `Params params, seen
+            | None ->
+              ( `Args (decl_params @ ret_args),
+                `Params (decl_params @ params),
+                seen )
+            | Some res -> (
+              (* Gather "existentials" (including types which are only bound because
+                 they're under another type expression) for the constructor.
+
+                 See step 2b and 2c from Note [With-bounds for GADTs] *)
+              let existentials =
+                !constructor_unbound_type_vars_excluding_row_variables cstr
+                |> Btype.TypeSet.to_seq
+                |> Seq.map Types.Transient_expr.type_expr
+                |> List.of_seq
+              in
+              let tof_kinds =
+                List.map
+                  (fun ty ->
+                    match get_desc ty with
+                    | Tof_kind _ ->
+                      (* We shouldn't be able to hit this, but it's harmless and
+                         defensive to just keep these types the same. *)
+                      ty
+                    | Tvar { jkind; _ } | Tunivar { jkind; _ } ->
+                      Btype.newgenty (Tof_kind jkind)
+                    | _ ->
+                      Misc.fatal_error
+                        "constructor_unbound_type_vars must return Tvar or \
+                         Tunivar")
+                  existentials
+              in
+              match Types.get_desc res with
+              | Tconstr (_, args, _) ->
+                List.fold_left2
+                  (fun (`Args ret_args, `Params params, seen) arg param ->
+                    if Btype.TypeSet.mem arg seen
+                    then
+                      (* We've already seen this type parameter, so don't add it again.
+                         See Step 2a from Note [With-bounds for GADTs] *)
+                      `Args ret_args, `Params params, seen
+                    else
+                      match Types.get_desc arg, Types.get_desc param with
+                      | Tvar _, Tvar _ ->
```

Comment:
The param technically doesn't have to be a var here.... but if it's not, a type error will be reported elsewhere. Test case:

```
type 'a t =
  K : 'b -> 'b t
  constraint 'a = int
```

Here, the arg will be a Tvar but the param will not be. I think it's just cleaner not to look at the param at all.